In [1]:
import warnings
from statsmodels.tools.sm_exceptions import ConvergenceWarning
warnings.simplefilter('ignore', ConvergenceWarning)

import os
import math
import numpy as np
import pandas as pd
from optbinning import OptimalBinning
from optbinning import BinningProcess
from optbinning.scorecard import plot_auc_roc, plot_cap, plot_ks
from sklearn.linear_model import LogisticRegression
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from scipy.stats.contingency import association
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report, auc, roc_auc_score, roc_curve, accuracy_score
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
from IPython.display import display

# Config the matlotlib backend as plotting inline in IPython
%matplotlib inline

In [2]:
# data = pd.read_csv('banking.csv', header = 0)
# data.to_hdf('banking.h5', key='data', mode='w') 
data = pd.read_hdf('banking.h5', 'data')
print(f"BR: {data['y'].mean():.6f}")
print(data.shape)
display(data)

BR: 0.112654
(41188, 21)


,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp_var_rate,cons_price_idx,cons_conf_idx,euribor3m,nr_employed,y
0,44,blue-collar,married,basic.4y,unknown,yes,no,cellular,aug,thu,...,1,999,0,nonexistent,1.4,93.444,-36.1,4.963,5228.1,0
1,53,technician,married,unknown,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-0.1,93.200,-42.0,4.021,5195.8,0
2,28,management,single,university.degree,no,yes,no,cellular,jun,thu,...,3,6,2,success,-1.7,94.055,-39.8,0.729,4991.6,1
3,39,services,married,high.school,no,no,no,cellular,apr,fri,...,2,999,0,nonexistent,-1.8,93.075,-47.1,1.405,5099.1,0
4,55,retired,married,basic.4y,no,yes,no,cellular,aug,fri,...,1,3,1,success,-2.9,92.201,-31.4,0.869,5076.2,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
41183,59,retired,married,high.school,unknown,no,yes,telephone,jun,thu,...,1,999,0,nonexistent,1.4,94.465,-41.8,4.866,5228.1,0
41184,31,housemaid,married,basic.4y,unknown,no,no,telephone,may,thu,...,2,999,0,nonexistent,1.1,93.994,-36.4,4.860,5191.0,0
41185,42,admin.,single,university.degree,unknown,yes,yes,telephone,may,wed,...,3,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,0
41186,48,technician,married,professional.course,no,no,yes,telephone,oct,tue,...,2,999,0,nonexistent,-3.4,92.431,-26.9,0.742,5017.5,0


In [3]:
train_data, test_data = train_test_split(data, test_size = 0.3, stratify = data.y, random_state=42)
print(train_data.shape)
print(test_data.shape)

(28831, 21)
(12357, 21)


#### Input variables

* age (numeric)
* job : type of job (categorical: “admin”, “blue-collar”, “entrepreneur”, “housemaid”, “management”, “retired”, “self-employed”, “services”, “student”, “technician”, “unemployed”, “unknown”)
* marital : marital status (categorical: “divorced”, “married”, “single”, “unknown”)
* education (categorical: “basic.4y”, “basic.6y”, “basic.9y”, “high.school”, “illiterate”, “professional.course”, “university.degree”, “unknown”)
* default: has credit in default? (categorical: “no”, “yes”, “unknown”)
* housing: has housing loan? (categorical: “no”, “yes”, “unknown”)
* loan: has personal loan? (categorical: “no”, “yes”, “unknown”)
* contact: contact communication type (categorical: “cellular”, “telephone”)
* month: last contact month of year (categorical: “jan”, “feb”, “mar”, …, “nov”, “dec”)
* day_of_week: last contact day of the week (categorical: “mon”, “tue”, “wed”, “thu”, “fri”)
* duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y=’no’). The duration is not known before a call is performed, also, after the end of the call, y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model
* campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
* pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
* previous: number of contacts performed before this campaign and for this client (numeric)
* poutcome: outcome of the previous marketing campaign (categorical: “failure”, “nonexistent”, “success”)
* emp.var.rate: employment variation rate — (numeric)
* cons.price.idx: consumer price index — (numeric)
* cons.conf.idx: consumer confidence index — (numeric)
* euribor3m: euribor 3 month rate — (numeric)
* nr.employed: number of employees — (numeric)

Predict variable (desired target):
* y — has the client subscribed a term deposit? (binary: “1”, means “Yes”, “0” means “No”)

In [4]:
target = 'y'

all_columns = train_data.columns.to_list()
all_columns.remove("nr_employed")

string_cols = train_data.select_dtypes(include='object').columns.tolist()
numeric_cols = train_data.select_dtypes(include='number').columns.tolist()

all_feats = [x for x in all_columns if ((x in numeric_cols + string_cols) and (x != target))]

In [5]:
binning_process = BinningProcess(variable_names = all_feats
                                 , max_n_prebins = 5
                                 , categorical_variables = string_cols
                                 , selection_criteria = None
                                 , n_jobs = os.cpu_count())

binning_process.fit(train_data[all_feats], train_data['y'].values)

summary = binning_process.summary()
display(summary)

,name,dtype,status,selected,n_bins,iv,js,gini,quality_score
0,age,numerical,OPTIMAL,True,2,0.105875,0.012662,0.088864,0.058751
1,job,categorical,OPTIMAL,True,5,0.192016,0.023185,0.217014,0.60588
2,marital,categorical,OPTIMAL,True,3,0.026001,0.003244,0.077227,0.041358
3,education,categorical,OPTIMAL,True,5,0.048307,0.00601,0.117602,0.091914
4,default,categorical,OPTIMAL,True,2,0.126471,0.015422,0.127014,0.220298
5,housing,categorical,OPTIMAL,True,2,0.000546,0.000068,0.011665,0.001195
6,loan,categorical,OPTIMAL,True,2,0.000316,0.00004,0.006719,0.000334
7,contact,categorical,OPTIMAL,True,2,0.247163,0.030252,0.219052,0.524563
8,month,categorical,OPTIMAL,True,5,0.401986,0.047718,0.302253,0.111373
9,day_of_week,categorical,OPTIMAL,True,5,0.006625,0.000828,0.042142,0.000655


In [6]:
train_data_woe = binning_process.transform(train_data, metric = "woe")
test_data_woe = binning_process.transform(test_data, metric = "woe")

df_train = train_data_woe.join(train_data[target], how='inner')
df_test = test_data_woe.join(test_data[target], how='inner')

In [7]:
import os
import random
import logging
import itertools
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.metrics import roc_auc_score
from scipy.stats.contingency import association
from concurrent.futures import ProcessPoolExecutor, as_completed
from functools import partial

logging.basicConfig(level=logging.WARNING, format="%(levelname)s: %(message)s")

# ------------------------------------------------------------------
# Configuration
# ------------------------------------------------------------------
max_workers = max(1, os.cpu_count() - 1)
all_features = list(all_feats)
min_n_feats = 2
max_n_feats = 10
max_n_models = 1000
target = "y"                                    # your target column

# Pre-select only needed columns once
cols_needed = all_features + [target]
df_train_slim = df_train[cols_needed].copy()
df_test_slim  = df_test[cols_needed].copy()

# ------------------------------------------------------------------
# Generate candidate models more carefully
# ------------------------------------------------------------------
n_feats = len(all_features)
all_models = []
for N in range(min_n_feats, min(max_n_feats, n_feats) + 1):
    combos = list(itertools.combinations(all_features, N))
    all_models.extend(combos)

print(f"{len(all_models):,} total combinations generated")

random.seed(42)                                   # reproducibility
random.shuffle(all_models)
all_models = all_models[:max_n_models]
print(f"{len(all_models)} models to evaluate")

# ------------------------------------------------------------------
# Worker function
# ------------------------------------------------------------------
def logistic_regression_model(included_features, df_train, df_test, target):
    """
    Fit a logistic regression model for a selected feature set and compute
    performance, coefficient, and multicollinearity diagnostics.
    """
    try:
        feats = list(included_features)
        X_train = df_train[feats]
        y_train = df_train[target].values
        X_test  = df_test[feats]
        n_cols  = len(feats)

        X_train_const = sm.add_constant(X_train, has_constant="add")
        model = sm.Logit(y_train, X_train_const).fit(disp=0, method="newton", maxiter=100)

        # Coefficients (including intercept)
        params = model.params
        pvalues = model.pvalues
        max_coef = params.max()
        max_p_value = pvalues.max()
        all_neg_betas = int(max_coef < 0)

        # Predictions
        proba_train = model.predict(X_train_const)
        proba_test  = model.predict(sm.add_constant(X_test, has_constant="add"))

        # Discriminatory power
        auc_train = roc_auc_score(y_train, proba_train)
        auc_test  = roc_auc_score(df_test[target], proba_test)
        gini_train = 2 * auc_train - 1
        gini_test  = 2 * auc_test - 1
        delta_gini = abs(gini_train - gini_test)

        # Multicollinearity
        X = X_train.values.astype(float)
        vif_list = [variance_inflation_factor(X, i) for i in range(X.shape[1])]
        max_vif = max(vif_list) if vif_list else np.nan

        # Pearson correlation (upper triangle)
        corr = np.corrcoef(X, rowvar=False)
        mask = np.triu(np.ones_like(corr, dtype=bool), k=1)
        max_pearson = np.abs(corr[mask]).max() if mask.any() else 0.0

        # Condition number (on column-normalised matrix)
        norms = np.linalg.norm(X, axis=0)
        norms[norms == 0] = 1.0          # safety
        X_norm = X / norms
        max_con_index = np.linalg.cond(X_norm)

        # Cramér's V correlation (upper triangle)
        corr = np.zeros((n_cols, n_cols))
        np.fill_diagonal(corr, 1.0)
        
        for i, col1 in enumerate(included_features):
            for j, col2 in enumerate(included_features):
                if i < j:
                    contingency_table = pd.crosstab(X_train[col1], X_train[col2])
                    cramer_v = association(contingency_table.values, method="cramer")
                    corr[i, j] = cramer_v
                    corr[j, i] = cramer_v
        
        mask = np.triu(np.ones_like(corr, dtype=bool), k=1)
        max_cramer_v = np.abs(corr[mask]).max() if mask.any() else 0.0

        return {
            "included_features": tuple(feats),
            "n_included_features": n_cols,
            "all_negative_betas": all_neg_betas,
            "max_pvalue": float(max_p_value),
            "gini_train": float(gini_train),
            "gini_test": float(gini_test),
            "delta_gini": float(delta_gini),
            "max_vif": float(max_vif),
            "max_con_index": float(max_con_index),
            "max_pearson": float(max_pearson),
            "max_cramer_v": float(max_cramer_v),
        }

    except Exception as e:
        logging.warning("Model %s failed: %s", included_features, e)
        return None

# ------------------------------------------------------------------
# Parallel execution
# ------------------------------------------------------------------
results = []
worker = partial(logistic_regression_model, df_train=df_train_slim, df_test=df_test_slim, target=target)

with ProcessPoolExecutor(max_workers=max_workers) as executor:
    futures = [executor.submit(worker, list(feats)) for feats in all_models]
    for future in as_completed(futures):
        res = future.result()
        if res is not None:
            results.append(res)

combined_df = pd.DataFrame(results)

354,502 total combinations generated
1000 models to evaluate


/home/krzysiek/Work/Python/.venv/lib/python3.12/site-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)
/home/krzysiek/Work/Python/.venv/lib/python3.12/site-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)
/home/krzysiek/Work/Python/.venv/lib/python3.12/site-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)
/home/krzysiek/Work/Python/.venv/lib/python3.12/site-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)
/home/krzysiek/Work/Python/.venv/lib/python3.12/site-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)
/home

In [8]:
# ------------------------------------------------------------------
# Post-processing
# ------------------------------------------------------------------
df_all_model = (
    combined_df
    .query(
        "all_negative_betas == 1 "
        "and max_pvalue < 0.05 "
        "and max_vif < 5 "
        "and max_pearson < 0.70 "
        "and max_cramer_v < 0.40"          # adjust threshold to portfolio
    )
    .sort_values("gini_test", ascending=False)
    .head(1000)
    .reset_index(drop=True)
)

display(df_all_model)
df_all_model.to_excel("all_models_output.xlsx", index=False)

,included_features,n_included_features,all_negative_betas,max_pvalue,gini_train,gini_test,delta_gini,max_vif,max_con_index,max_pearson,max_cramer_v
0,"(job, default, contact, duration, previous)",5,1,1.911958e-30,0.750112,0.756252,0.006140,1.107695,1.457826,0.241173,0.241173
1,"(job, default, duration, previous)",4,1,1.432625e-38,0.734731,0.739018,0.004287,1.040960,1.248610,0.121330,0.183381
2,"(age, marital, education, duration, poutcome)",5,1,1.628001e-19,0.723725,0.729822,0.006097,1.038492,1.257672,0.126449,0.132486
3,"(age, education, default, contact, day_of_week...",6,1,6.264759e-03,0.720471,0.723520,0.003049,1.046032,1.275205,0.138039,0.158464
4,"(marital, education, emp_var_rate)",3,1,1.709652e-02,0.497716,0.495273,0.002442,1.029682,1.191706,0.122118,0.099378
5,"(job, education, default, campaign, poutcome, ...",6,1,1.924579e-05,0.504300,0.494025,0.010275,1.237607,1.721356,0.346021,0.346406
6,"(age, marital, default, campaign, poutcome, co...",6,1,1.550275e-09,0.498757,0.489771,0.008986,1.252952,1.657502,0.346021,0.346406
7,"(contact, previous)",2,1,9.766236e-71,0.339917,0.331920,0.007997,1.078525,1.318743,0.241173,0.241173
